fix(QueryContext): validation does not validate query_context metrics

ofekisr opened the pull request:
### SUMMARY

close 
continue in #19753

amitmiran137 reviewed (changes requested):
Extract until you can't

amitmiran137 reviewed (approved):
LGTM !
Feature introduced here will enforce datasource access on custom SQL expression of metrics 

@villebro we should add column in a follow up PR

craig-rueda reviewed (changes requested):
Thanks for this PR - it's bumping up security! I don't see any tests, however. Also, there's a lot of single line methods in the ContextValidator. Can we compress those a bit? Too much delegation makes the code a bit hard to follow along with.

jayakrishnankk commented on superset/charts/api.py:
```diff
@@ -617,7 +619,7 @@ def get_data(self, pk: int) -> Response:
 
         try:
-            command = ChartDataCommand()
+            command = self._create_chart_data_command()
```
What's left TO DO in this, so that it can get merged ?

ofekisr replied:
@jayakrishnankk review and merge

ofekisr committed:
fix query_context validation does not validate sql metrics

ofekisr committed:
remove mark and clean small issue in query_context.py

ofekisr commented:
@craig-rueda @hughhhh @jayakrishnankk 
After all the recent refactor in ChartData, I recommitted the PR with all the adjustments to refactored code

villebro commented on superset/security/manager.py:
```diff
@@ -1047,17 +1048,23 @@ def raise_for_access(
-                )
+            self.raise_when_there_is_no_access_to(datasource)
+
+    def raise_when_there_is_no_access_to(self, datasource: BaseDatasource) -> None:
```
This name should IMO follow the existing naming conventions. In this case `raise_for_datasource_access` would be more appropriate and more aligned with the pre-existing `raise_for_dashboard_access`.

ofekisr replied:
"raise_for_dashboard_acess" is even more bad name. what is mean "dashboard_acess" ?

a better name conventions are on of those: validate_<object_type>_can_be_accessed
or 
actor_can_access_<object_type>

betodealmeida replied:
We can discuss which naming strategy is better, but I think we can all agree that consistency is a good thing. If we have an existing pattern I think it would be better to stick with it, since it makes it easier for people to understand the code.

etr2460 replied:
noting that this pattern was originally used (i think) because it follows the similar one used by the python requests package, `raise_for_status`: https://docs.python-requests.org/en/latest/api/#requests.Response.raise_for_status

@john-bodley might know a bit more about why we picked this naming

ofekisr replied:
@etr2460 
You gave a good reference, but I still think is an anti-pattern, and they are wrong too.
but with that reference, it should be discussed in a different context not in that PR

@betodealmeida 
There is a conflict between two principles: 
consistency vs you should always leave the code you have just modified cleaner than it was, a good catch and currently, I don't have a good argument or opinion on what principle is stronger. 

I'll rename it

villebro commented on tests/common/query_context_generator.py:
```diff
@@ -248,12 +302,77 @@ def generate(
+                "isNew": False,
+                "label": "COUNT({})".format(column_name),
+                "optionName": "metric_bvvzfjgdg7_eawlsdp84tq",
+                "sqlExpression": "(SELECT count(name) FROM {})".format(table_name),
```
nit: 
```suggestion
                "sqlExpression": f"(SELECT count(name) FROM {table_name})",
```

villebro commented on tests/common/query_context_generator.py:
```diff
@@ -248,12 +302,77 @@ def generate(
+                "expressionType": "SQL",
+                "hasCustomLabel": False,
+                "isNew": False,
+                "label": "COUNT({})".format(column_name),
```
nit:
```suggestion
                "label": f"COUNT({column_name})",
```

villebro commented on superset/charts/data/query_context_validator.py:
```diff
@@ -0,0 +1,168 @@
+        else:
+            self._security_manager.raise_for_access(query_context=query_context)
+
+    def _validate_when_context_based_on_sql_database(
```
This has come up in previous review comments from me. While I agree that it's good practice to break out logical chunks to improve readability and enforce separation of concerns, there is a point when abstraction can actually decrease readability. Each nested call causes additional cognitive overhead for the reviewer, and dilutes the relevance of other more important functions/methods. As a rule of thumb, if a private method is 1) only used once 2) consists of less than 5 lines of code 3) further branches out into another private method which also satisfies 1) and 2), I would consider just inlining the code and adding a comment describing what is happening (if needed).

craig-rueda replied:
Yep, I agree. Too much delegation makes the logic flow hard to follow

villebro commented on superset/security/manager.py:
```diff
@@ -974,14 +976,14 @@ def set_perm(  # pylint: disable=unused-argument
-        datasource: Optional["BaseDatasource"] = None,
-        query: Optional["Query"] = None,
-        query_context: Optional["QueryContext"] = None,
-        table: Optional["Table"] = None,
-        viz: Optional["BaseViz"] = None,
+        database: Optional[Database] = None,
+        datasource: Optional[BaseDatasource] = None,
+        query: Optional[Query] = None,
+        query_context: Optional[QueryContext] = None,
+        table: Optional[Table] = None,
+        viz: Optional[BaseViz] = None,
```
Is this nowadays the correct way of referring to imports from `TYPE_CHECKING`?

ofekisr replied:
a string representation is worst than the real type

betodealmeida replied:
If the imports are inside an `if TYPE_CHECKING` then you need to use a string, otherwise you get a `NameError`.

ofekisr replied:
> 
When you use the old method of type hint as string, you actually don't need to import the class at all and don't need to put it under TYPE_CHECKING
but if you would like to use the new better choice for marking the type hint by the real type 
you should import them under TYPE_CHECKING with the __future__ annotations.
That is the better choice so you can navigate and let the IDEs do their job --> assist you

villebro reviewed (commented):
First pass comments. I think this is a great improvement to the security model, but I'm concerned this may cause false positives (I did a similar change some time ago, which caused unpleasant regressions due to triggering false positives in edge cases that hadn't been thought of in the original PR). To make rolling out new security improvements less problematic, I would almost consider introducing (yet another) feature flag "SECURITY_BETA" where new security improvements can be introduced and kept until more comprehensive testing has been done.

amitmiran137 committed:
fix: string format best practice

Co-authored-by: Ville Brofeldt <33317356+villebro@users.noreply.github.com>

amitmiran137 committed:
fix: string format best practice

Co-authored-by: Ville Brofeldt <33317356+villebro@users.noreply.github.com>

ofekisr commented on superset/charts/data/commands/get_data_command.py:
```diff
@@ -14,6 +14,8 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+from __future__ import annotations
+
```
I develop with py 3.8, without it raises errors

craig-rueda replied:
Yep, let's remove this if we can.

ofekisr replied:
https://docs.python.org/3/library/__future__.html

![image](https://user-images.githubusercontent.com/35701650/144020171-d5ecaa5e-c8df-4aac-8499-8455b071f98f.png)

ofekisr replied:
@villebro 
regarding the FF, I don't agree with it since the change is fixing a bug a not adding new functionality.
if someone will encounter a surprising issue with it, it means "hey it got something, we have a security breach" so the app administrator should think about how to solve it, giving extra permission or removing the metric.

this is my IMO, but for now I'll add that FF

villebro replied:
@ofekisr My point for introducing the FF is to gradually introduce the new stricter security check in a way which makes it possible to catch potential regressions that weren't caught during initial review. By having an introductory period during which the stricter checks are behind a FF, orgs that require the new stricter checking can help harden the feature by pushing fixes to master before the new checks are made generally available.

ofekisr commented on superset/security/manager.py:
```diff
@@ -14,8 +14,10 @@
 # specific language governing permissions and limitations
 # under the License.
-# pylint: disable=too-many-lines
+# pylint: disable=too-many-lines, invalid-name
```
I don't think we should use the invalid-name rule as well... 
I use the clean code principles, so there is no sense to put ignoring on each statement

betodealmeida replied:
Putting on each statement means you're aware of the rule, and you consciously decided to break it because practicality beats purity. If you put it at the top of the model the rule could be broken many times without people being aware of it, which reduces the code quality.

john-bodley replied:
+1. The desired pattern is mentioned [here](https://github.com/apache/superset/blob/master/CONTRIBUTING.md#python).

ofekisr replied:
I'll add and open new discussion on it

ofekisr committed:
fix pr requested changes
  - put new validation under FF
  - consistency with naming convention
  - remove global pylint disable

ofekisr committed:
fix failing unit tests

ofekisr committed:
fix pr requested change

ofekisr committed:
fix pre-commit and pylint

jayakrishnankk commented:
@craig-rueda - requesting review. bumping this.

villebro commented on tests/common/query_context_generator.py:
```diff
@@ -199,16 +199,16 @@ def get_query_object(
-            **QUERY_OBJECTS[parent_query_name],
-            **obj,
-        }
+        obj = QUERY_OBJECTS[query_name]
```
I don't understand the point of this change - if `query_name not in QUERY_OBJECTS`, we probably can't assign `obj = QUERY_OBJECTS[query_name]`? Consequently, the code below becomes unreachable, raising the question if it's needed anymore

ofekisr replied:
this is not the updated code

ofekisr committed:
add: integration tests

ofekisr committed:
fix failed tests

ofekisr committed:
Merge branch 'upstream_master' into fix/sql_metric_access_validation

# Conflicts:
#	tests/integration_tests/charts/data/api_tests.py

ofekisr committed:
fix failed tests

ofekisr committed:
fix failed tests

ofekisr committed:
fix failed tests

ofekisr committed:
fix failed tests
